### DISCUSSION TITLE
Impersonation and programmatical login

### DISCUSSION THREAD

**Discussion by igor-alexandrov** [janko/rodauth-rails#236] [Q&A] [answered]
Hello,

I want to provide an ability for admins to impersonate an existing account.
I created a separate configuration for this.

``` ruby
# frozen_string_literal: true

class RodauthImpersonate < RodauthBase
  configure do
    enable :login, :logout, :internal_request

    session_key_prefix "impersonate_"

    # ==> Redirects
    # Redirect to home page after logout.
    logout_redirect '/'

    # Redirect user to their company page after login. Works only if `login_return_to_requested_location_path` is nil.
    login_redirect { rails_routes.company_path(rails_account.member.company) }
  end
end
```

And call it like this: 

``` ruby
def impersonate
    rodauth = Rodauth::Rails.rodauth(:impersonate, account: Member.find(params[:id]).account)
    rodauth.login('impersonate')

    redirect_to company_url(resource_member.company)
end
```

I have no errors, but it simply does not work.

Do you have any ideas on implementing this and what I am doing wrong?

Thank you.

**Comment by janko** [comment]
Yeah, this example won't work, because `Rodauth::Rails.rodauth` goes through the `internal_request` feature, which doesn't have any connection to the current request, so it cannot update the session.

I don't know what's the best way to implement this. My gut tells me that a separate configuration isn't ideal, because then in places where you're calling e.g. `rodauth.require_authentication` you also need to call `rodauth(:impersonate).require_authentication`. I think it's better to use the main configuration for impersonation, assuming admins have their own separate configuration & sessions in your app. Something like:

```rb
def impersonate
  rodauth(:admin).require_account # only admins can impersonate accounts

  account = Member.find(params[:id]).account

  rodauth.account_from_login(account.email)
  rodauth.login("impersonate")
  # if main configuration is using multifactor authentication
  rodauth.two_factor_update_session("impersonate-two") if rodauth.uses_two_factor_authentication?

  redirect_to company_url(resource_member.company)
end
```

And then in your layout you could add a banner or indication notifying the admin that they're impersonating another user:

```erb
<% if rodauth(:admin).authenticated? && rodauth.authenticated? %>
  <div>You're impersonating user <%= rodauth.rails_account.email %></div>
<% end %>
```

When logging out from impersonated account, you just need to be careful to avoid clearing the admin session as well, because by default the whole session is cleared. I don't remember now what's the recommended way to do that.

**Reply by igor-alexandrov** [reply]
Thank you for the clarification. I will implement it and come back to you.

**Comment by igor-alexandrov** [comment]
Sorry for disturbing you again.

I decided to combine my original idea with yours. So I still have a separate configuration to impersonate users. It still seems to me, that it would be easier, but again I faced problems.

This works:

``` ruby
account = Member.find(params[:id]).account

rodauth.account_from_login(account.email)
rodauth.login("impersonate")
# the line below will never be executed, because `login` makes a redirect. Probably better to use `login_session`
rodauth.two_factor_update_session("impersonate-2fa") if rodauth(:impersonate).uses_two_factor_authentication?
```

However this does not work and `rodauth(:impersonate).rails_account` is `nil`.

``` ruby
account = Member.find(params[:id]).account

rodauth(:impersonate).account_from_login(account.email)
rodauth(:impersonate).login_session("impersonate")
rodauth(:impersonate).two_factor_update_session("impersonate-2fa") if rodauth(:impersonate).uses_two_factor_authentication?
```

Don't really have any ideas where is the problem.

**Reply by janko** [reply]
The likely problem is that the `:impersonate` configuration stores the account ID in `session[:impersonate_account_id]` (due to the session prefix), while the primary configuration presumably has no session prefix, in which case it looks for `session[:account_id]`. This would mean that accounts logged in through the `:impersonate` configuration will never be found by the main configuration.

**Reply by janko** [reply]
My mistake, didn't realize you were talking about `rodauth(:impersonate).rails_account` returning nil (and not `rodauth.rails_account`). Can you tell me what SQL queries the `rodauth(:impersonate).rails_account` call makes, or the `rodauth(:impersonate).require_account` call if you're using that?

**Reply by igor-alexandrov** [reply]
As of now, `rodauth(:impersonate).rails_account` makes no SQL queries, which probably means there is no session.

Forget to say, I added `session_key_prefix` to the `RodauthImpersonate` configuration.

``` ruby
class RodauthImpersonate < RodauthBase
  configure do
    enable :login, :logout

    rails_controller { RodauthController }

    session_key_prefix "impersonate_"

    # ==> Redirects
    # Redirect to home page after logout.
    logout_redirect '/'

    login_redirect nil
  end
end
```

**Reply by janko** [reply]
For me it works when I add it to the official demo app, `rodauth(:impersonate).rails_account` returns the logged in account.

```diff
diff --git a/app/controllers/admin/rodauth_controller.rb b/app/controllers/admin/rodauth_controller.rb
index b5e04d0..0d54a77 100644
--- a/app/controllers/admin/rodauth_controller.rb
+++ b/app/controllers/admin/rodauth_controller.rb
@@ -1,2 +1,12 @@
 class Admin::RodauthController < Admin::ApplicationController
+  def impersonate
+    rodauth(:admin).require_account
+
+    account = Account.where(type: "main").first
+
+    rodauth(:impersonate).account_from_login(account.email)
+    rodauth(:impersonate).login_session("impersonate")
+    # rodauth(:impersonate).two_factor_update_session("impersonate-2fa") if rodauth(:impersonate).uses_two_factor_authentication?
+    redirect_to root_path
+  end
 end
diff --git a/app/misc/rodauth_app.rb b/app/misc/rodauth_app.rb
index d9209be..0627931 100644
--- a/app/misc/rodauth_app.rb
+++ b/app/misc/rodauth_app.rb
@@ -1,6 +1,7 @@
 class RodauthApp < Rodauth::Rails::App
   configure RodauthMain
   configure RodauthAdmin, :admin
+  configure RodauthImpersonate, :impersonate
 
   route do |r|
     rodauth.load_memory # autologin remembered users
diff --git a/app/misc/rodauth_impersonate.rb b/app/misc/rodauth_impersonate.rb
new file mode 100644
index 0000000..45b2e58
--- /dev/null
+++ b/app/misc/rodauth_impersonate.rb
@@ -0,0 +1,21 @@
+class RodauthImpersonate < RodauthBase
+  configure do
+    enable :login, :logout
+
+    rails_controller { RodauthController }
+
+    session_key_prefix "impersonate_"
+
+    # ==> Redirects
+    # Redirect to home page after logout.
+    logout_redirect '/'
+
+    login_redirect nil
+  end
+
+  private
+
+  def account_type
+    "main"
+  end
+end
diff --git a/app/views/home/index.html.erb b/app/views/home/index.html.erb
index c6e531b..562d697 100644
--- a/app/views/home/index.html.erb
+++ b/app/views/home/index.html.erb
@@ -5,3 +5,5 @@
   <p class="lead">This application demonstrates a possible integration of <%= link_to "Rodauth", "http://rodauth.jeremyevans.net", target: "_blank" %> into a Rails application, using the <%= link_to "rodauth-rails", "https://github.com/janko/rodauth-rails", target: "_blank" %> gem.</p>
   <p class="lead">You can find the source code at <%= link_to "github.com/janko/rodauth-demo-rails", "https://github.com/janko/rodauth-demo-rails", target: "_blank" %>.</p>
 </div>
+
+<%= rodauth(:impersonate).rails_account.inspect %>
diff --git a/config/routes.rb b/config/routes.rb
index d33e301..6914857 100644
--- a/config/routes.rb
+++ b/config/routes.rb
@@ -4,5 +4,6 @@ Rails.application.routes.draw do
 
   namespace :admin do
     root to: "home#index"
+    get "/impersonate" => "rodauth#impersonate"
   end
 end
```

**Reply by igor-alexandrov** [reply]
Okay, I cloned the demo app, and I found two things:

1. You cannot have two active sessions at the same time. To reproduce this, sign in to admin, then follow `/admin/impersonate` path, and you will get logged in as an impersonated user, but your admin session will no longer exist.

2. It seems I also reproduced my original bug: sign in as a regular user, then go to `/admin` and try to sign in as admin. You will see that based on SQL queries, you have been logged in, but you are getting redirected to `/admin/login`. This cannot be reproduced if you first sign in to `/admin` and, after this, will try to sign in as a regular user.

**Reply by igor-alexandrov** [reply]
My original idea was to have two sessions at the same time, with different identifiers. I 100% sure, that Authlogic allowed to do this.

**Reply by igor-alexandrov** [reply]
Okay, this happens because on each login, `clear_session` method is called, and it does not care about prefixes or something. It just clears the whole session values.

Seems like it should clear only the keys related to the configuration that initiated the login.

**Reply by igor-alexandrov** [reply, marked as answer]
Ok, I found a solution. As I written above `clear_session` deleted all session keys, no matter what configuration created them.
My main configuration had `remeber` feature enabled. So here is what happened:
1. You click on "impersonate" button
2. Controller action logged in impersonated user
3. `clear_session` was called, and removed all session keys
4. `remember` feature from the main configuration checked that session is not valid (it didn't exist) and restored it and called `clear_session` again, which cleared impersonated user session.

Don't know is it a bug or not, but for now I fixed this by redefining `clear_session`.

``` ruby
# frozen_string_literal: true

class RodauthImpersonate < Rodauth::Rails::Auth
  configure do
    enable :login, :logout

    prefix '/impersonate'

    session_key_prefix 'impersonate_'

    logout_redirect '/'
    # logout_notice_flash nil

    login_redirect nil
    login_route nil
    # login_notice_flash nil
  end

  def clear_session
    session.keys.each do |key|
      session.delete(key) if key.start_with?(session_key_prefix)
    end
  end
end
```

**Reply by janko** [reply]
Ah, that makes sense. It seems that the main problem is Rodauth resetting the session regardless of whether other configurations have their session data as well. While gems like Devise/Warden store session data in a sub-hash within a single top-level session key, Rodauth stores all session data on the top level, and doesn't currently store a list of its session keys.

I considered adding the ability to Rodauth for storing session keys in a sub-hash, and in general allowing multiple configurations to be logged in at the same time, but I never started. I would suggest proposing this to Jeremy and see what he thinks, as I remember others getting bit by this as well in the past.

**Reply by igor-alexandrov** [reply]
Thanks @janko I will ask Jeremy.

**Comment by pboling** [comment]
This impersonate feature is very similar to the become_account feature.

https://github.com/adam12/rodauth-become_account/issues/3
